packages/salesforcedx-vscode-apex/src/views/testOutline.ts

```diff
@@ -0,0 +1,666 @@
+import { ApexTestRequestInfo, getApexTests, isLanguageClientReady } from '..';
+import { nls } from '../messages';
+
+// Import from core
+const sfdxCoreExports = vscode.extensions.getExtension(
+  'salesforce.salesforcedx-vscode-core'
+)!.exports;
+const ProgressNotification = sfdxCoreExports.ProgressNotification;
+const EmptyParametersGatherer = sfdxCoreExports.EmptyParametersGatherer;
+const SfdxCommandlet = sfdxCoreExports.SfdxCommandlet;
+const ForceApexTestRunCodeActionExecutor =
+  sfdxCoreExports.ForceApexTestRunCodeActionExecutor;
+const SfdxWorkspaceChecker = sfdxCoreExports.SfdxWorkspaceChecker;
+const channelService = sfdxCoreExports.channelService;
+const notificationService = sfdxCoreExports.notificationService;
+const taskViewService = sfdxCoreExports.taskViewService;
+
+const LIGHT_BLUE_BUTTON = ospath.join(
+  __filename,
+  '..',
+  '..',
+  '..',
+  '..',
+  'resources',
+  'light',
+  'testNotRun.svg'
+);
+const LIGHT_RED_BUTTON = ospath.join(
+  __filename,
+  '..',
+  '..',
+  '..',
+  '..',
+  'resources',
+  'light',
+  'testFail.svg'
+);
+const LIGHT_GREEN_BUTTON = ospath.join(
+  __filename,
+  '..',
+  '..',
+  '..',
+  '..',
+  'resources',
+  'light',
+  'testPass.svg'
+);
+
+const DARK_BLUE_BUTTON = ospath.join(
+  __filename,
+  '..',
+  '..',
+  '..',
+  '..',
+  'resources',
+  'dark',
+  'testNotRun.svg'
+);
+const DARK_RED_BUTTON = ospath.join(
+  __filename,
+  '..',
+  '..',
+  '..',
+  '..',
+  'resources',
+  'dark',
+  'testFail.svg'
+);
+const DARK_GREEN_BUTTON = ospath.join(
+  __filename,
+  '..',
+  '..',
+  '..',
+  '..',
+  'resources',
+  'dark',
+  'testPass.svg'
+);
+
+export type FullTestResult = {
+  summary: TestSummary;
+  tests: TestResult[];
+};
+
+export type TestSummary = {
+  outcome: string;
+  testsRan: number;
+  passing: number;
+  failing: number;
+  skipped: number;
+  passRate: string;
+  failRate: string;
+  testStartTime: string;
+  testExecutionTime: string;
+  testTotalTime: string;
+  commandTime: string;
+  hostname: string;
+  orgId: string;
+  username: string;
+  testRunId: string;
+  userId: string;
+};
+
+export type TestResult = {
+  ApexClass: ApexClass;
+  MethodName: string;
+  Outcome: string;
+  RunTime: number;
+  Message: string;
+  StackTrace: string;
+  FullName: string;
+};
+
+export type ApexClass = {
+  attributes: { type: string };
+  Id: string;
+  Name: string;
+  NamespacPrefix: string;
+};
+
+export type ApexGroupInfo = {
+  outcome: string;
+  summary: TestSummary;
+};
+
+export class ApexTestOutlineProvider implements vscode.TreeDataProvider<Test> {
+  private _onDidChangeTreeData: vscode.EventEmitter<
+    Test | undefined
+  > = new vscode.EventEmitter<Test | undefined>();
+  public readonly onDidChangeTreeData: vscode.Event<Test | undefined> = this
+    ._onDidChangeTreeData.event;
+
+  private eventsEmitter = new events.EventEmitter();
+
+  private apexTestMap: Map<string, Test> = new Map<string, Test>();
+  private head: Test | null;
```

Comment:
Pick a better name than `head`.

Reply:
Well, it's the object that contains all the Test Classes and tests, so '<container>'?

Reply:
Is this the rootNode for all child tree items (tests, test classes, etc)?
Maybe `rootNode` or `treeRoot`?